### PR TITLE
ijsdk-491

### PR DESCRIPTION
**Pull request by JohnHake**
New content about UI Theme APIs
* Adds content about creating/editing UI Themes
* Adds supporting content about the Internal Actions menu: UI inspector and LaF defaults.

/cc @YannCebron  for review of new internal actions docs

### CODE REVIEW

File: reference_guide/ui_themes/themes_extras.md [hunk 1]
```diff
@@ -37,7 +37,7 @@ Users can do this manually in [Preferences](https://www.jetbrains.com/help/idea/
 
 UI Themes support specifying a background image as a key-value pair in the `"background": {}` section of a Theme description file:
 * The `image` key uses the file name of the image as the value.
-The background image must be in PNG format, and placed in the UI Theme plugin project's resources folder. 
+The background image should be in PNG format, and placed in the UI Theme plugin project's resources folder.
```
Comment (YannCebron) [inline reference_guide/ui_themes/themes_extras.md]: why "should be in PNG"?

File: reference_guide/internal_actions/enabling_internal.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+* From the main menu, select **Help | Edit Custom Properties**.
+This selection opens IntelliJ IDEA's `idea.properties` file.
+If a properties file does not exist, IntelliJ IDEA will prompt to create one.
+* Add the line shown below to the `idea.properties` file:
+![Enable Internal Mode](img/internal_mode_enable.png) 
+* Save the `idea.properties` file and restart IntelliJ IDEA.
+
+The Internal Actions menu is available in **Tools \| Internal Actions**.
+
+![Internal Actions](img/internal_action_menu.png)
```
Comment (YannCebron) [inline reference_guide/internal_actions/enabling_internal.md]: remove screenshot, adds no value IMHO

File: reference_guide/internal_actions/enabling_internal.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+The Internal Actions menu is only visible if internal mode is enabled in IntelliJ IDEA.
+
+## Enabling Internal Mode
+There are multiple ways to enable internal mode, but the simplest is within IntelliJ IDEA:
+* Start IntelliJ IDEA.
+* From the main menu, select **Help | Edit Custom Properties**.
+This selection opens IntelliJ IDEA's `idea.properties` file.
+If a properties file does not exist, IntelliJ IDEA will prompt to create one.
+* Add the line shown below to the `idea.properties` file:
+![Enable Internal Mode](img/internal_mode_enable.png)
```
Comment (YannCebron) [inline reference_guide/internal_actions/enabling_internal.md]: remove screenshot, adds no value IMHO
provide contents to be pasted in docs as text for c/p

File: reference_guide/internal_actions/enabling_internal.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+---
+
+The Internal Actions menu is only visible if internal mode is enabled in IntelliJ IDEA.
+
+## Enabling Internal Mode
+There are multiple ways to enable internal mode, but the simplest is within IntelliJ IDEA:
+* Start IntelliJ IDEA.
+* From the main menu, select **Help | Edit Custom Properties**.
+This selection opens IntelliJ IDEA's `idea.properties` file.
+If a properties file does not exist, IntelliJ IDEA will prompt to create one.
```
Comment (YannCebron) [inline reference_guide/internal_actions/enabling_internal.md]: "If it does not exist, ..."

File: reference_guide/internal_actions/internal_ui_sub.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
+---
+title: Internal Actions - UI Submenu
+---
+
+The Internal Actions UI submenu provides JetBrains IDE plugin developers with a suite of tools to help develop, debug, and test their IntelliJ Platform project UI.
```
Comment (YannCebron) [inline reference_guide/internal_actions/internal_ui_sub.md]: -"JetBrains IDE"

File: reference_guide/internal_actions/internal_uii.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+If the menu item **Tools \| Internal Actions \| UI \| UI Inspector** is not available in IntelliJ IDEA, then the first step is to [enable internal mode](enabling_internal.md)
+
+## Enabling the UI Inspector
+Before using the _UI Inspector_, it must be enabled by selecting the menu item **Tools \| Internal Actions \| UI \| UI Inspector**.
+The enabled state of the _UI Inspector_ is modal; it remains enabled until it is disabled by selecting the _UI Inspector_ menu item again. 
+
+When enabled, the _UI Inspector_ menu entry displays a check mark as shown below:
+
+![Enabled UI Inspector Menu](img/internal_uii_menu.png)
```
Comment (YannCebron) [inline reference_guide/internal_actions/internal_uii.md]: remove screenshot

File: reference_guide/internal_actions/internal_uii.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+## Using the UI Inspector
+While enabled, centering the cursor on a UI element and pressing _control-alt_ when _clicking_ the mouse reveals the internal description of the UI element.
+
+For example, to get information about the build (hammer) icon on the toolbar, put the cursor on the icon and press _control-alt_ while clicking the mouse.
+As shown below:
```
Comment (YannCebron) [inline reference_guide/internal_actions/internal_uii.md]: missing linebreak

File: reference_guide/internal_actions/internal_ui_lafd.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+The _LaF Defaults_ window provides a key-value pair lookup for UI Controls.
+It also allows interactive prototyping of UI Control color changes. 
+
+If the menu item **Tools \| Internal Actions** is not available in IntelliJ IDEA, then the first step is to [enable internal mode](enabling_internal.md)
+
+## Opening LaF Defaults
+The _LaF Defaults_ window is opened by selecting the menu item **Tools \| Internal Actions \| UI \| LaF Defaults**.
+
+The _LaF Defaults_ window has two columns representing key-value pairs for UI Controls: 
+* The _Name_ column contains the UI Control `key` for each IntelliJ IDEA UI element available at runtime.
```
Comment (YannCebron) [inline reference_guide/internal_actions/internal_ui_lafd.md]: - "IntelliJ IDEA"

File: reference_guide/internal_actions/internal_ui_lafd.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+* The _Value_ column contains the UI Control color `value` for each IntelliJ IDEA UI element.
```
Comment (YannCebron) [inline reference_guide/internal_actions/internal_ui_lafd.md]: - "IntelliJ IDEA"

File: reference_guide/ui_themes/themes.md [hunk 10]
```diff
@@ -0,0 +1,163 @@
+---
+title: Custom UI Themes
+---
+
+Custom UI Themes give designers control of the appearance of built-in IntelliJ IDEA UI elements. Custom 
+UI Themes are distinct from [creating new UI elements](/user_interface_components/user_interface_components.md)
+for a plugin. 
+
+Creating a custom UI Theme is a process of choosing a base IntelliJ Theme (_Light_ or _Darcula_) 
+then changing aspects of the base Theme definition. Custom UI Themes can:
```
Comment (YannCebron) [inline reference_guide/ui_themes/themes.md]: linebreak

File: reference_guide/ui_themes/themes.md [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+* An optional Editor Scheme description (XML) file derived from an exported IntelliJ editor scheme.
+This file is declared in the Theme description file and located in a plugin project's resources folder. 
+
+(Screenshot image of project window highlighting each component)
+
+## Custom UI Theme Workflow
+Creating a UI Theme follows this general sequence:
+* [Start with a Plugin Project](https://www.jetbrains.org/intellij/sdk/docs/basics/getting_started/creating_plugin_project.html)
+* [Add UI Theme components to the plugin](#creating-a-ui-theme-with-the-devkit-theme-wizard)
+  by using the DevKit UI Theme Wizard.
```
Comment (YannCebron) [inline reference_guide/ui_themes/themes.md]: DevKit --> "Plugin DevKit" (more occurrences below)

File: reference_guide/ui_themes/themes.md [hunk 12]
```diff
@@ -40,26 +41,28 @@ Creating a UI Theme follows this general sequence:
...
 ### Creating a UI Theme with the DevKit Theme Wizard
-To add a UI Theme to a plugin use the DevKit Wizard. While a plugin project is 
-open in IntelliJ, select the _**New | Plugin DevKit | Theme**_ action. This Wizard can
-be used for both DevKit and Gradle-based plugins.
+The DevKit Wizard is part of the DevKit plugin, which is bundled and enabled in IntelliJ IDEA. 
+While a plugin project is open in IntelliJ IDEA, select the _**New | Plugin DevKit | Theme**_ action. 
+This Wizard can be used for both DevKit and Gradle-based plugins.
 
-(Image of menu selecting plugin wizard.)
+![DevKit Wizard Action](img/devkit_wiz_action.png)
```
Comment (YannCebron) [inline reference_guide/ui_themes/themes.md]: remove screenshot

File: reference_guide/ui_themes/themes.md [hunk 13]
```diff
@@ -40,26 +41,28 @@ Creating a UI Theme follows this general sequence:
...
+
+<br>
 
 The Wizard then prompts for the name of the new Theme, and the basis for the Theme:
 
-(Image of Create New Theme dialog box)
+![DevKit Dialog](img/devkit_wiz_dialog.png)
```
Comment (YannCebron) [inline reference_guide/ui_themes/themes.md]: remove screenshot

File: reference_guide/ui_themes/themes.md [hunk 14]
```diff
@@ -40,26 +41,28 @@ Creating a UI Theme follows this general sequence:
...
 
-A UI Theme component is
-added to a plugin using the DevKit UI Theme Wizard.
+A UI Theme is added to an IntelliJ Platform plugin using the DevKit UI Theme Wizard.
 
 ### Creating a UI Theme with the DevKit Theme Wizard
-To add a UI Theme to a plugin use the DevKit Wizard. While a plugin project is 
-open in IntelliJ, select the _**New | Plugin DevKit | Theme**_ action. This Wizard can
-be used for both DevKit and Gradle-based plugins.
+The DevKit Wizard is part of the DevKit plugin, which is bundled and enabled in IntelliJ IDEA.
```
Comment (YannCebron) [inline reference_guide/ui_themes/themes.md]: - "and enabled"

File: reference_guide/ui_themes/themes_customize.md [hunk 15]
```diff
@@ -195,11 +173,17 @@ The following example sets a new border width and color for all windows:
...
 any global color override.
 
-### Finding UI Control Attributes
-There are hundreds of UI control `control.property` keys in the IntelliJ UI. Unfortunately, 
-there isn't autocomplete functionality in the editor to assist programmers. Some
-keys and strategies for applying them can be gleaned from the
-[UI Theme reference implementations](#UI-Theme-Reference-Implementations). 
+## Finding Attribute Keys for UI Controls
+There are hundreds of UI control `element.property` keys defined in the IntelliJ IDEA UI. 
+Unfortunately, there isn't autocomplete functionality in the IDE Editor to assist programmers.
```
Comment (YannCebron) [inline reference_guide/ui_themes/themes_customize.md]: ???

File: reference_guide/ui_themes/themes_customize.md [hunk 16]
```diff
@@ -195,11 +173,17 @@ The following example sets a new border width and color for all windows:
 In this example the customized border supersedes the default definition and 
 any global color override.
 
-### Finding UI Control Attributes
-There are hundreds of UI control `control.property` keys in the IntelliJ UI. Unfortunately, 
-there isn't autocomplete functionality in the editor to assist programmers. Some
-keys and strategies for applying them can be gleaned from the
-[UI Theme reference implementations](#UI-Theme-Reference-Implementations). 
+## Finding Attribute Keys for UI Controls
+There are hundreds of UI control `element.property` keys defined in the IntelliJ IDEA UI.
```
Comment (YannCebron) [inline reference_guide/ui_themes/themes_customize.md]: IntelliJ Platform UI

File: reference_guide/ui_themes/themes_extras.md [hunk 17]
```diff
@@ -2,28 +2,24 @@
...
 
-UI Themes can also provide custom color and font settings, as well as custom 
-images for display in the IntelliJ application window.
+UI Themes can also provide custom color and font settings, as well as custom images for display in the IntelliJ IDEA application window.
  
 ## Adding a Custom Editor Scheme
-IntelliJ users can set preferences to configure the colors and fonts used in the Editor. 
-These custom color and font settings are called _color schemes_, and can be 
-exported and shared.
+IntelliJ IDEA users can set preferences to configure the colors and fonts used in the Editor.
```
Comment (YannCebron) [inline reference_guide/ui_themes/themes_extras.md]: IDE

File: reference_guide/ui_themes/themes_extras.md [hunk 18]
```diff
@@ -2,28 +2,24 @@
...
+IntelliJ IDEA users can set preferences to configure the colors and fonts used in the Editor. 
+These custom color and font settings are called _color schemes_, and can be exported and shared.
 
-Use the following procedure to create a color scheme for a UI Theme. If needed refer to the
-[Configuring Colors and Fonts](https://www.jetbrains.com/help/idea/configuring-colors-and-fonts.html)
-section of IntelliJ help:
-* Create the desired custom color scheme using the IntelliJ preferences.
+Use the following procedure to create a color scheme for a UI Theme. 
+If needed refer to the [Configuring Colors and Fonts](https://www.jetbrains.com/help/idea/configuring-colors-and-fonts.html) section of IntelliJ IDEA help:
+* Create the desired custom color scheme using the IntelliJ IDEA preferences.
```
Comment (YannCebron) [inline reference_guide/ui_themes/themes_extras.md]: IDE preferences

File: reference_guide/ui_themes/themes_extras.md [hunk 19]
```diff
@@ -36,21 +32,17 @@ custom UI Theme:
 ```
 
 ## Adding a Custom Background Image
-IntelliJ supports setting an image as a background in the application window. 
-Users can do this manually in 
-[Preferences](https://www.jetbrains.com/help/idea/setting-background-image.html).
+IntelliJ IDEA supports setting an image as a background in the application window.
```
Comment (YannCebron) [inline reference_guide/ui_themes/themes_extras.md]: The IDE

File: reference_guide/ui_themes/themes_customize.md [hunk 20]
```diff
@@ -175,14 +175,17 @@ In this example the customized border supersedes the default definition and
...
-Unfortunately, there isn't autocomplete functionality in the IDE Editor to assist programmers. 
+There are hundreds of UI control `element.property` keys defined in the IntelliJ Platform UI. 
 Some keys and strategies for applying them can be gleaned from the [UI Theme reference implementations](#UI-Theme-Reference-Implementations). 
-
 For a general search, here some suggested methods for locating UI control keys.
 
+### Finding a UI Control Key Using Code Completion in the Editor
+The preferred method of finding UI control keys is to use the code completion feature in the IntelliJ IDEA editor.
+New entries in the `"ui": {}` section will invoke the code completion popup, as shown below:
+![UI Control Key Code Completion](img/uit_control_complete.png)
```
Comment (YannCebron) [inline reference_guide/ui_themes/themes_customize.md]: there's a linebreak issue here with the image

File: reference_guide/ui_themes/themes_customize.md [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+The syntax of a Theme description file follows the JSON open-standard file format of key-value pairs. 
+The minimum valid default file is the output of the [DevKit Theme Wizard](themes.md#creating-a-ui-theme-with-the-devkit-theme-wizard). 
+Adding key-value pairs customizes a Theme.
+
+UI Theme key-value pairs often use a color as the `value`.
+Colors are defined by six-digit RGB or eight-digit RGBA hexadecimal notation.
+
+### UI Theme Reference Implementations
+When learning new syntax, it is often useful to have some existing implementations for reference.
+For example, refer to the [Theme description file](upsource:///platform/platform-resources/src/themes/HighContrast.theme.json) for the IntelliJ IDEA _High Contrast_ Theme.
```
Comment (OlyaB) [inline reference_guide/ui_themes/themes_customize.md]: Not sure how this work but when I clicked this link in the prebuild website sent by Yann, an outdated version of the file was opened https://www.dropbox.com/s/zu0i23tpkt6ch52/Screenshot%202019-01-23%2014.33.38.png?dl=0

File: reference_guide/internal_actions/internal_uii.md [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+
+When enabled, the _UI Inspector_ menu entry displays a check mark.
+
+## Using the UI Inspector
+While enabled, centering the cursor on a UI element and pressing _control-alt_ when _clicking_ the mouse reveals the internal description of the UI element.
+
+For example, to get information about the build (hammer) icon on the toolbar, put the cursor on the icon and press _control-alt_ while clicking the mouse.
+As shown below, the _UI Inspector_ displays:
+* The icon has the internal path `Allcons.Actions.Compile`.
+* The `foreground` and `background` colors for the icon are also displayed.
```
Comment (OlyaB) [inline reference_guide/internal_actions/internal_uii.md]: I think mentioning "foreground and background" icon colors might be misleading — these colors do not have any effect on the icon itself. There are no Icon.background or Icon.foreground keys. These colors are from the panel containing the icon and can be changed with Panel.background and Panel.foreground keys. https://www.dropbox.com/s/0sbalzsj2ss9cbn/Screenshot%202019-01-23%2014.54.36.png?dl=0

File: reference_guide/ui_themes/themes_customize.md (hunk 21 above)
Comment (YannCebron) [reply reference_guide/ui_themes/themes_customize.md]: we're always linking to latest release, so that is 183.4284 (2018.3)

**Pull request merged**